### DISCUSSION TITLE
Issue with enabling user supplied id in JanusGraph

### DISCUSSION THREAD

**Discussion by lesleyDeng** [JanusGraph/janusgraph#4660] [Q&A] [answered]
Hello,

I am encountering an issue while trying to configure JanusGraph to allow user-supplied-ids. I followed the guide and made the necessary settings, but it seems they are not taking effect. Here are the details of my environment and the steps I have taken:
JanusGraph version: 1.0.0
I started the Gremlin console using ./bin/gremlin.sh.
Following the documentation found at [Custom Vertex ID](https://docs.janusgraph.org/master/advanced-topics/custom-vertex-id/)， I executed the following commands:
`
gremlin> graph = JanusGraphFactory.open('inmemory')

15:50:12 INFO  org.janusgraph.diskstorage.configuration.builder.ReadConfigurationBuilder.setupTimestampProvider - Set default timestamp provider MICRO

15:50:12 INFO  org.janusgraph.graphdb.idmanagement.UniqueInstanceIdRetriever.getOrGenerateUniqueInstanceId - Generated unique-instance-id=7f0001012947944-ubuntu1

15:50:12 INFO  org.janusgraph.diskstorage.configuration.ExecutorServiceBuilder.buildFixedExecutorService - Initiated fixed thread pool of size 320

15:50:12 INFO  org.janusgraph.graphdb.database.StandardJanusGraph.<init> - Gremlin script evaluation is disabled

15:50:12 INFO  org.janusgraph.diskstorage.log.kcvs.KCVSLog$MessagePuller.initializeTimepoint - Loaded unidentified ReadMarker start time 2024-08-29T07:50:12.927784Z into org.janusgraph.diskstorage.log.kcvs.KCVSLog$MessagePuller@148fca83

==>standardjanusgraph[inmemory:[127.0.0.1]]

gremlin> mgmt = graph.openManagement()

==>org.janusgraph.graphdb.database.management.ManagementSystem@3a13f663

gremlin> mgmt.set("graph.set-vertex-id", true)

==>org.janusgraph.diskstorage.configuration.UserModifiableConfiguration@56f61d74

gremlin> mgmt.commit()

==>null
`

However, when I called graph.features(), the result indicates that the setting has not been applied.  See
`
gremlin> graph.features()

==>FEATURES

> VertexFeatures
...
>-- UserSuppliedIds: false

`

I am aware that the setting **graph.set-vertex-id** is of **GLOBAL_OFFLINE** mutability level according to the [Configuration Reference](https://docs.janusgraph.org/configs/configuration-reference/), so I have not started any traversal instances on the graph, which I believe is the correct approach.

Could you please advise what might be causing this issue and how I can resolve it?

Thank you for your assistance.

Best regards,
Lesley

**Comment by li-boxuan** [comment]
What you have observed is expected. You will see 

> VertexFeatures
...
-- UserSuppliedIds: true

once you enable `graph.set-vertex-id`

**Reply by lesleyDeng** [reply]
Yes, I was expecting to see that result. Currently, it's showing as UserSuppliedIds: false.

**Reply by li-boxuan** [reply, marked as answer]
Oh sorry I read your post too quickly. IIRC  in-memory backends don't support changing GLOBAL_OFFLINE configs because you basically need to restart your graph for new configs to take effect. There's no "restart" concept for in-memory graphs since they are not persistent.

You could try to set the values when you start the graph.

**Reply by lesleyDeng** [reply]
Thank you, I think your point is very reasonable, 'restart' and 'in-memory' seem like conflicting concepts. I just tried your suggestion, as follows, and running graph.features() shows that it has taken effect (very much appreciated for your professional advice). 

`gremlin> graph = JanusGraphFactory.open("/pathto/janusgraph-inmemory.properties")` 

janusgraph-inmemory.properties:
`
gremlin.graph=org.janusgraph.core.JanusGraphFactory
storage.backend=inmemory
graph.set-vertex-id=true
`
and I have
` 
-- UserSuppliedIds: true
...
-- NumericIds: true
`

However, when I tried to add a vertex with a specified ID,  I encountered a new error: 'Not a valid vertex id: 10'. e.g., 
`
gremlin> g = graph.traversal()
gremlin> g.addV('person').property(id,10)
Not a valid vertex id: 10
`

The source code seems to be from the line where the error is reported: 
`Preconditions.checkArgument(vertexId == null || IDManager.VertexIDType.NormalVertex.is(vertexId), "Not a valid vertex id: %s", vertexId);`. 
The ID I provided should be supported as NumericIds: true.

**Reply by lesleyDeng** [reply]
Using `g.addV('person').property(id, graph.getIDManager().toVertexId(10))` and `graph.getIDManager().fromVertexId()` can solve my problem
